InputOutput.cpp

```diff
@@ -59,4 +65,53 @@ void InputOutput<Container>::display_container() const {
   for (const auto &pair : container) {
     std::cout << pair.first << " " << pair.second << std::endl;
   }
+}
+
+template <template <typename...> typename Container>
+void InputOutput<Container>::open_file_to_write(
+    const std::string &algorithm_context) {

```

Comment:
IO object has reference to Context  object and the same Context object has reference to the same IO object? or its impossible to happend? if possible - change this logic, its bad practice

Reply:
Context object has a reference to IO object and not the other way round and the context can call IO's methods.

After the algorithm was run in context, here in line 72 we pass just the context name (a field of `AlgorithmContext` - maybe the parameter `algorithm context` should be renamed) so that `AlgorithmContext` can call it's `io` field's method `process_output_file` and IO can create a proper output file name.